File: 2015/day05.py
def vowel_count(s: str) -> int:
    total = 0
    for c in s:
        if c in ['a', 'e', 'i', 'o', 'u']:
            total += 1
    return total

def double_letter(s: str) -> bool:
    for c in range(1, len(s)):
        if s[c] == s[c-1]:
            return True

def part_one(strings: [str]) -> int:
    nice = 0
    for string in strings:
        bad = False
        for x in ["ab", "cd", "pq", "xy"]:
            if x in string:
                bad = True
        if bad:
            continue
        if vowel_count(string) >= 3 and double_letter(string):
            nice += 1
    return nice

def double_double(s: str) -> bool:
    for i in range(len(s) - 3):
        chunk = s[i:i+2]
        rest = s[i+2:]
        if chunk in rest:
            return True

def seperated_double(s: str) -> bool:
    for c in range(2, len(s)):
        if s[c] == s[c-2]:
            return True

def part_two(strings: [str]) -> int:
    nice = 0
    for string in strings:
        if double_double(string) and seperated_double(string):
            nice += 1
    return nice

with open("data/day05.txt", "r") as file:
    strings = file.read().splitlines()
    print(part_one(strings))
    print(part_two(strings))

File: 2015/Scala/day05.scala
import scala.io.Source
import scala.util.Using

object day05 {

  def partOne(strings: List[String]): Int = {
    def stringIsNice(s: String): Boolean = {
      val vowels = s.count(Set('a', 'e', 'i', 'o', 'u').contains)
      val doubleLetter = s.sliding(2).exists(_.distinct.length == 1)
      val forbidden = s.contains("ab") || s.contains("cd") || s.contains("pq") || s.contains("xy")
      vowels >= 3 && doubleLetter && !forbidden
    }

    strings.count(stringIsNice)
  }

  def partTwo(strings: List[String]): Int = {
    def stringIsNice(s: String): Boolean = {
      val doublePair = s.indices.dropRight(1).exists { i =>
        val pair = s.substring(i, i + 2)
        s.indexOf(pair, i + 2) != -1
      }
      val repeatLetter = s.sliding(3).exists(chunk => chunk(0) == chunk(2))
      doublePair && repeatLetter
    }

    strings.count(stringIsNice)
  }

  def main(args: Array[String]): Unit = {
    val inputFile = "2015/data/day05.txt"

    val strings = Using.resource(Source.fromFile(inputFile)) { source =>
      source.getLines().toList
    }

    println(partOne(strings))
    println(partTwo(strings))
  }

}
